<a href="https://colab.research.google.com/github/letter-b/lab-feature-engineering/blob/main/lab_feature_eng.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# LAB | Feature Engineering

**Load the data**

In this challenge, we will be working with the same Spaceship Titanic data, like the previous Lab. The data can be found here:

https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv

Metadata

https://github.com/data-bootcamp-v4/data/blob/main/spaceship_titanic.md

In [22]:
#Libraries
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor

In [2]:
spaceship = pd.read_csv("https://raw.githubusercontent.com/data-bootcamp-v4/data/main/spaceship_titanic.csv")
spaceship.head()

,PassengerId,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Name,Transported
0,0001_01,Europa,False,B/0/P,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,Maham Ofracculy,False
1,0002_01,Earth,False,F/0/S,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,Juanna Vines,True
2,0003_01,Europa,False,A/0/S,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,Altark Susent,False
3,0003_02,Europa,False,A/0/S,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,Solam Susent,False
4,0004_01,Earth,False,F/1/S,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,Willy Santantines,True


**Check the shape of your data**

In [3]:
spaceship.shape

(8693, 14)

**Check for data types**

In [4]:
spaceship.dtypes

,0
PassengerId,object
HomePlanet,object
CryoSleep,object
Cabin,object
Destination,object
Age,float64
VIP,object
RoomService,float64
FoodCourt,float64
ShoppingMall,float64


**Check for missing values**

In [5]:
spaceship.isna().sum()

,0
PassengerId,0
HomePlanet,201
CryoSleep,217
Cabin,199
Destination,182
Age,179
VIP,203
RoomService,181
FoodCourt,183
ShoppingMall,208


There are multiple strategies to handle missing data

- Removing all rows or all columns containing missing data.
- Filling all missing values with a value (mean in continuos or mode in categorical for example).
- Filling all missing values with an algorithm.

For this exercise, because we have such low amount of null values, we will drop rows containing any missing value.

In [6]:
spaceship = spaceship.dropna()

- **Cabin** is too granular - transform it in order to obtain {'A', 'B', 'C', 'D', 'E', 'F', 'G', 'T'}

In [8]:
spaceship["Cabin"].head()

spaceship["Cabin"] = spaceship["Cabin"].str[0]
spaceship["Cabin"].head()

,Cabin
0,B
1,F
2,A
3,A
4,F


- Drop PassengerId and Name

In [12]:
columns_to_drop = [col for col in ["PassengerId", "Name"] if col in spaceship.columns]
if columns_to_drop:
    spaceship = spaceship.drop(columns_to_drop, axis=1)
spaceship.head()

,HomePlanet,CryoSleep,Cabin,Destination,Age,VIP,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Transported
0,Europa,False,B,TRAPPIST-1e,39.0,False,0.0,0.0,0.0,0.0,0.0,False
1,Earth,False,F,TRAPPIST-1e,24.0,False,109.0,9.0,25.0,549.0,44.0,True
2,Europa,False,A,TRAPPIST-1e,58.0,True,43.0,3576.0,0.0,6715.0,49.0,False
3,Europa,False,A,TRAPPIST-1e,33.0,False,0.0,1283.0,371.0,3329.0,193.0,False
4,Earth,False,F,TRAPPIST-1e,16.0,False,303.0,70.0,151.0,565.0,2.0,True


- For non-numerical columns, do dummies.

In [13]:
spaceship = pd.get_dummies(spaceship)
spaceship.head()

,Age,RoomService,FoodCourt,ShoppingMall,Spa,VRDeck,Transported,HomePlanet_Earth,HomePlanet_Europa,HomePlanet_Mars,...,Cabin_D,Cabin_E,Cabin_F,Cabin_G,Cabin_T,Destination_55 Cancri e,Destination_PSO J318.5-22,Destination_TRAPPIST-1e,VIP_False,VIP_True
0,39.0,0.0,0.0,0.0,0.0,0.0,False,False,True,False,...,False,False,False,False,False,False,False,True,True,False
1,24.0,109.0,9.0,25.0,549.0,44.0,True,True,False,False,...,False,False,True,False,False,False,False,True,True,False
2,58.0,43.0,3576.0,0.0,6715.0,49.0,False,False,True,False,...,False,False,False,False,False,False,False,True,False,True
3,33.0,0.0,1283.0,371.0,3329.0,193.0,False,False,True,False,...,False,False,False,False,False,False,False,True,True,False
4,16.0,303.0,70.0,151.0,565.0,2.0,True,True,False,False,...,False,False,True,False,False,False,False,True,True,False


**Perform Train Test Split**

In [18]:
X = spaceship.drop('Transported', axis=1)
y = spaceship['Transported']

In [19]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state=0)

**Model Selection**

In this exercise we will be using **KNN** as our predictive model.

In [24]:
knn = KNeighborsRegressor(n_neighbors=10)

- Evaluate your model's performance. Comment it

In [27]:
knn.fit(X_train, y_train)

KNeighborsRegressor(n_neighbors=10)

In [28]:
print(f"The R2 of the model on the TEST set is: {knn.score(X_test, y_test): .2f}")

The R2 of the model on the TEST set is:  0.40


An R² of 0.40 means that approximately 40% of the variance in whether a passenger is 'Transported' can be explained by the model's features.

Below is me testing things out.

Evaluating with Classification Metrics

Since the target variable `Transported` is binary (True/False), this is a classification problem. While `KNeighborsRegressor` attempts to predict a continuous value (often between 0 and 1 in this context), we can convert these predictions into binary classes (e.g., if the prediction is >= 0.5, classify as True; otherwise, classify as False) and then use standard classification metrics.

In [29]:
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# Make predictions on the test set
y_pred_proba = knn.predict(X_test)

# Convert probabilities to binary predictions (0 or 1)
# A common threshold is 0.5, but this can be tuned.
y_pred = (y_pred_proba >= 0.5).astype(bool)

# Calculate classification metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print(f"Accuracy: {accuracy:.2f}")
print(f"Precision: {precision:.2f}")
print(f"Recall: {recall:.2f}")
print(f"F1-Score: {f1:.2f}")

Accuracy: 0.78
Precision: 0.76
Recall: 0.83
F1-Score: 0.79


These scores give us a much clearer picture of the model's performance on this classification task compared to the R² score.


* An Accuracy of 0.78 means that the model correctly predicted whether a passenger was 'Transported' or not about 78% of the time.
* A Precision of 0.76 indicates that when the model predicts a passenger was 'Transported', it was correct 76% of the time.
* A Recall of 0.83 suggests that the model successfully identified 83% of all passengers who were actually 'Transported'.
* The F1-Score of 0.79 is a balanced measure, considering both precision and recall. It's a good overall indicator of the model's performance.